src/main/flight/navigation_rewrite_multicopter.c

```diff
@@ -86,12 +86,15 @@ static void updateAltitudeVelocityController_MC(uint32_t deltaMicros)
     float altitudeError = posControl.desiredState.pos.V.Z - posControl.actualState.pos.V.Z;
     float targetVel = altitudeError * posControl.pids.pos[Z].param.kP;
 
-    // hard limit desired target velocity to +/- 20 m/s
-    targetVel = constrainf(targetVel, -2000.0f, 2000.0f);
-
-    // limit max vertical acceleration 250 cm/s/s - reach the max 20 m/s target in 80 seconds
-    float maxVelDifference = US2S(deltaMicros) * 250.0f;
-    posControl.desiredState.vel.V.Z = constrainf(targetVel, posControl.desiredState.vel.V.Z - maxVelDifference, posControl.desiredState.vel.V.Z + maxVelDifference);
+    // hard limit desired target velocity to max_climb_rate
+    targetVel = constrainf(targetVel, -posControl.navConfig->max_climb_rate, posControl.navConfig->max_climb_rate);
+
+    // limit max vertical acceleration to 1/5G (~200 cm/s/s) if we are increasing velocity.
+    // if we are decelerating - don't limit (allow better recovery from falling)
+    if (ABS(targetVel) > ABS(posControl.desiredState.vel.V.Z)) {
+        float maxVelDifference = US2S(deltaMicros) * (GRAVITY_CMSS / 5.0f);
+        posControl.desiredState.vel.V.Z = constrainf(targetVel, posControl.desiredState.vel.V.Z - maxVelDifference, posControl.desiredState.vel.V.Z + maxVelDifference);
+    }
```

Comment:
Thanks! I've missed that bug!

On Mon, May 30, 2016 at 11:46 PM, theArchLadder <notifications@github.com>
wrote:

> In src/main/flight/navigation_rewrite_multicopter.c
> <https://github.com/iNavFlight/inav/pull/257#discussion_r65105238>:
>
> > @@ -86,12 +86,15 @@ static void updateAltitudeVelocityController_MC(uint32_t deltaMicros)
> >      float altitudeError = posControl.desiredState.pos.V.Z - posControl.actualState.pos.V.Z;
> >      float targetVel = altitudeError * posControl.pids.pos[Z].param.kP;
> >
> > -    // hard limit desired target velocity to +/- 20 m/s
> > -    targetVel = constrainf(targetVel, -2000.0f, 2000.0f);
> > -
> > -    // limit max vertical acceleration 250 cm/s/s - reach the max 20 m/s target in 80 seconds
> > -    float maxVelDifference = US2S(deltaMicros) * 250.0f;
> > -    posControl.desiredState.vel.V.Z = constrainf(targetVel, posControl.desiredState.vel.V.Z - maxVelDifference, posControl.desiredState.vel.V.Z + maxVelDifference);
> > +    // hard limit desired target velocity to max_climb_rate
> > +    targetVel = constrainf(targetVel, -posControl.navConfig->max_climb_rate, posControl.navConfig->max_climb_rate);
> > +
> > +    // limit max vertical acceleration to 1/5G (~200 cm/s/s) if we are increasing velocity.
> > +    // if we are decelerating - don't limit (allow better recovery from falling)
> > +    if (ABS(targetVel) > ABS(posControl.desiredState.vel.V.Z)) {
> > +        float maxVelDifference = US2S(deltaMicros) * (GRAVITY_CMSS / 5.0f);
> > +        posControl.desiredState.vel.V.Z = constrainf(targetVel, posControl.desiredState.vel.V.Z - maxVelDifference, posControl.desiredState.vel.V.Z + maxVelDifference);
> > +    }
>
> Am i an idiot or is posControl.desiredState.vel.V.Z never set if target
> velocity is slowing down...?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/iNavFlight/inav/pull/257/files/cf255613fcb0e8626b5aa82d33d356ee9a3db6da#r65105238>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKi_m0Tw-jCLjem68ConzVcYscVopY2Tks5qG0y0gaJpZM4Ip9m4>
> .
>